### PR TITLE
Add viewpoint GET and xml->json mapper

### PR DESCRIPTION
**Pull request by oliverguenther**
- XML -> JSON Converter
  - [x] Converter implementation
  - [x] Converter tests
  - [x] Create migration for viewpoints => JSON
  - [x] Create circle test JSON -> XML -> JSON equality
  - [x] Write XML mapper
  - [x] Fix xml tests
  - [x] Use XML mapper in bcf-xml export

- [x] GET `<topic>/viewpoints`
- [x] GET `<topic>/viewpoints/<uuid>`
- [x] GET`<topic>/viewpoints/<uuid>/snapshot`
- [x] GET`<topic>/viewpoints/<uuid>/selection`
- [x] GET`<topic>/viewpoints/<uuid>/coloring`
- [x] GET`<topic>/viewpoints/<uuid>/visibility`

- [x] Use JSON mapper in BCF-XML import
- [x] Output raw json (subpaths) in viewpoints API

**Out of scope**
- POST `<topic>/viewpoints`
- Bitmaps
  - GET`<topic>/viewpoints/<uuid>/bitmaps`
  - GET`<topic>/viewpoints/<uuid>/bitmaps/<uuid>`
- (both are responded to with 501)

### CODE REVIEW

**Comment by oliverguenther** [comment]
I moved the factory bot definitions into test _and_ dev group to make it easier to access them for testing /using on the rails console

File: .codeclimate.yml [hunk 1]
```diff
@@ -27,6 +27,9 @@ checks:
 plugins:
   rubocop:
     enabled: true
+    checks:
+      Rubocop/Naming/ClassAndModuleCamelCase:
```
Comment (oliverguenther) [inline .codeclimate.yml:31]: This disables the codeclimate warning for `BCF::V2_1` due to the underscore

File: lib/api/errors/not_implemented.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See docs/COPYRIGHT.rdoc for more details.
+#++
+
+module API
+  module Errors
+    class NotImplemented < ErrorBase
+      identifier 'urn:openproject-org:api:v3:errors:NotImplemented'.freeze
```
Comment (oliverguenther) [inline lib/api/errors/not_implemented.rb]: The identifier is V3 specific but we reuse the errors, maybe we should prepend the identifiers in some APIv3 base class instead?

Comment (ulferts) [reply lib/api/errors/not_implemented.rb]: We could also remove the api v3 specific part from here and simply prepend it again in the api v3 error representer (e.g. https://github.com/opf/openproject/blob/dev/lib/api/v3/errors/error_representer.rb#L59).

File: modules/bcf/lib/open_project/bcf/bcf_json/viewpoint_mapper.rb [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        # Hoist exceptions components up from the nested XML node
+        exceptions = visibility.dig('exceptions', 'component')
+        visibility['exceptions'] = Array(exceptions) if exceptions
+
+        # Move view_setup_hints
+        view_setup_hints = hash.dig('components', 'view_setup_hints')
+        visibility['view_setup_hints'] = view_setup_hints.transform_values { |val| val == 'true' }
+
+        # Remove the old node
+        hash['components'].delete('view_setup_hints')
```
Comment (ulferts) [inline modules/bcf/lib/open_project/bcf/bcf_json/viewpoint_mapper.rb]: The mapper currently does not map the `index` property (see https://github.com/buildingSMART/BCF-API/blob/release_2_1/Schemas_draft-03/Collaboration/Viewpoint/viewpoint_GET.json)

Comment (ulferts) [reply modules/bcf/lib/open_project/bcf/bcf_json/viewpoint_mapper.rb]: Same is true for bitmaps. I am aware, that the end point is not supported, which is fine with me, but we should at least have the property.

**Review by ulferts — DISMISSED** [review]
Mostly looking good. Please see my tiny remarks.

File: modules/bcf/app/controllers/bcf/api/v2_1/topics_api.rb [hunk 4]
```diff
@@ -56,6 +56,8 @@ def topics
               .new(model: Bcf::Issue,
                    api_name: 'Topics')
               .mount
+
+        mount Bcf::API::V2_1::ViewpointsAPI
```
Comment (wielinde) [inline modules/bcf/app/controllers/bcf/api/v2_1/topics_api.rb]: @ulferts Curiosity: Why did you decide to put the API below `app/controllers/...` and not into `lib/api/...` as the BCF XML API?

File: modules/bcf/lib/open_project/bcf/bcf_json/viewpoint_mapper.rb [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+      ##
+      # Retrieve the viewpoint hash without root node, if any.
+      def viewpoint_hash
+        @viewpoint_hash ||= begin
+          # Load from XML using activesupport
+          hash = Hash.from_xml(resource.viewpoint)
+          hash = hash[ROOT_NODE] if hash.key?(ROOT_NODE)
+
+          # Perform destructive transformations
+          transformations.each do |m|
```
Comment (wielinde) [inline modules/bcf/lib/open_project/bcf/bcf_json/viewpoint_mapper.rb]: I would prefer something like `method_name` over `m`. Just one thing less to think of.

**Review by wielinde — APPROVED** [review]
That was a very interesting review for me. Nice!

I bet, we need to see that code in action, to find the little bugs.

I did not see any obvious issues with the existing BCF XML logic. However, and that is unrelated to the PR, we still need to put uniqueness constrains on the UUID fields.

File: modules/bcf/lib/open_project/bcf/bcf_json/viewpoint_mapper.rb (hunk 5 above)
Comment (oliverguenther) [reply modules/bcf/lib/open_project/bcf/bcf_json/viewpoint_mapper.rb]: Done 👍

**Pull request merged**